### ISSUE TITLE
alerts fail to evaluate queries and expressions

### ISSUE THREAD

**Issue by mhristof** [grafana/opensearch-datasource#173]
**What happened**:
when trying preview an alert, im getting the following error
```
Failed to evaluate queries and expressions: [plugin.downstreamError] failed to query data: Failed to query data: rpc error: code = Unknown desc = error while Decoding to MultiSearchResponse: EOF
```

![image](https://github.com/grafana/opensearch-datasource/assets/24205462/6b29cbe8-c967-4a25-afdd-c94eb3097fd8)
![Screenshot 2023-05-16 at 14 49 51](https://github.com/grafana/opensearch-datasource/assets/24205462/2e90f388-2b86-4917-a7ed-5895e287d604)

**What you expected to happen**:
The alert to preview the values

**How to reproduce it (as minimally and precisely as possible)**:
Create an alert as above and try to preview the status

**Anything else we need to know?**:

**Environment**:
- Grafana version: v9.5.2 (cfcea75916)
- Plugin version: 2.4.1

**Comment by sarahzinger** [comment]
Hi @mhristof thanks for opening this issue! 

I'm having some trouble recreating this but I think my test dataset is just different and it's hard to match the query to catch this edge case. Would you mind showing us a screenshot of what the query results look like in table format when on the query editor/explore page with any sensitive data removed so we might try to recreate this on our end?

Thanks so much!

**Comment by mhristof** [comment]
i think you mean this

![image](https://github.com/grafana/opensearch-datasource/assets/24205462/48747037-e49b-4174-b08e-39ddcc533216)

The query returns lots of empty messages like this

![image](https://github.com/grafana/opensearch-datasource/assets/24205462/4d013a5b-5387-4076-942e-00788b42afa8)

and the rows with values look like this

![image](https://github.com/grafana/opensearch-datasource/assets/24205462/c2722e51-dd45-4b0a-a2be-208df43946a0)

**Comment by sarahzinger** [comment]
@mhristof that's very helpful thanks! Still haven't quite recreated it yet, but I wonder if the issue is related to the missing values. 

I see we have this additional option for missing values where they can be set to a specific value such as zero:
![Screenshot 2023-05-17 at 10 37 17 AM](https://github.com/grafana/opensearch-datasource/assets/6620164/361e9753-8b1d-40a8-8df7-f372c073ef46)

Does altering that number help?

**Comment by sarahzinger** [comment]
It's not the same error message as what you're seeing but I am seeing inconsistent data in general between alerts and queries. Made another issue to track that here: https://github.com/grafana/opensearch-datasource/issues/176

**Comment by sarahzinger** [comment]
I'm so sorry that fix didn't work for you! I think it's likely we have some kind of bug in how our plugin is handling alerting queries, particularly around nil values, but even more generally I'm seeing inconsistent results between the results in our query editor and the results in our alerts. This is likely because unlike many of our other datasource plugins, this one seems to have 2 different query paths, one for alerting and one for the query editor. In our other plugins, we tend to have the same data flow path for both, which ensures any bug fixes in one area are fixed in the other.

We're currently scoping out a refactor of this plugin to unify the experience between alerting queries and query editor queries, and it's one of our top priorities for the quarter. But unfortunately it may take us at least a few weeks before we can ship improvements, though we will do what we can to expedite this process!

I am curious has alerting with this datasource worked well for you in the past and recently started showing errors? Just want to rule out that this isn't related to some sort of recent changes we may have made on our end. I suspect not and that these issues have likely been around for a long time, but would like to rule that out in case there's a quick fix, we can get out to unblock you.

**Comment by sarahzinger** [comment]
Great thanks @mhristof will do! cc @fridgepoet and @idastambuk for your refactor project!

**Comment by maxwellvarner** [comment]
I was previously able to create alerts with this plugin using the following versions. 

Grafana: 9.4.2
Opensearch Plugin: 2.1.0

I recently upgraded to Opensearch Plugin 2.4.1 and Grafana 9.5.2. I upgraded the plugin first, but don't recall testing to see if I could still create alerts. I can say that as of today with both recent upgrades I've done; I cannot create alerts and get the same error message as seen in the original post.

**Comment by pavriet-boxtal** [comment]
Hello, I have a related issue that I hope can help us get to the root of the problem.

Everything is fine while creating the alert, but a few times a day, we receive a notification for a datasourceError, and the Error annotation says `plugin.downstreamError] failed to query data: Failed to query data: rpc error: code = Unknown desc = error while Decoding to MultiSearchResponse: invalid character '}' looking for beginning of object key string`.

At first we were running v2.4.1, I upgraded to 2.6.2 after seeing this thread but nothing changed.

grafana v9.5.1 (bc353e4b2d)
opensearch-datasource 2.6.2
Datasource setup :
![01](https://github.com/grafana/opensearch-datasource/assets/136579429/db176273-5d1c-4d7a-a5af-4a6e0fac9866)

Hope this helps :)

**Comment by idastambuk** [comment]
Hi @pavriet-boxtal, thanks for reporting this. I just have a few questions: 
1. Can you tell us more about the query you are running for the alert?
2. Does the Version field  in the config form have the correct version for your OpenSearch instance? If not, can you run "Get version and Save"?

Thank you!

**Comment by pavriet-boxtal** [comment]
Hi @idastambuk,

1. Really not much to say imo
![image](https://github.com/grafana/opensearch-datasource/assets/136579429/514e4169-cf5e-4b53-9e1c-667427c93a21)

In the picture bellow, I was able to collect the same query (in a panel editor this time) where an error happened (blue vertical line @16:07) + the query inspect menu on the right. Just to show that data does not seem to be the issue.
![image](https://github.com/grafana/opensearch-datasource/assets/136579429/2a919e08-532a-4771-b83b-e9b69fdf5439)

2. For the full story, we use **opensearch 1.3** from AWS opensearch service. Before upgrading to v2.6.2 of the plugin, the Version field in the data source config form was set to opensearch 1.x (iirc). The 'Get version and save' button was not present. We had the errors in that state of config. 
Now, with v2.6.2, I saw the 'Get version and save', clicked it and it returned ES 7.10.2, as we see in the picture in my first comment. We still have the error, and we can not change the version field (weird behavior imo).

**Comment by fridgepoet** [comment]
Hi @pavriet-boxtal, thanks for all the information so far. What version of the plugin were you using when you did not have any issues?

**Comment by pavriet-boxtal** [comment]
Hello @fridgepoet , We had issues with both 2.4.1 and 2.6.2. We never had no issues :(

**Comment by idastambuk** [comment]
Hi @pavriet-boxtal, can you by any chance recall if the error for 2.4.1 was the same as in 2.6.2 (`error while Decoding to MultiSearchResponse: invalid character '}' looking for beginning of object key string`)? 
I'm just trying to determine if both versions have the same bug, or if 2.4.1 had the `error while Decoding to MultiSearchResponse: EOF`  and 2.6.1 is a different one. In which case we will track it in another ticket. Thanks!

**Comment by pavriet-boxtal** [comment]
I search our notification channels for EOF and found this (url anonymized for obvious reasons). I check and it happened twice while on **2.4.1**. I also attached a picture of the alert query that generated the errors.

`[plugin.downstreamError] failed to query data: Failed to query data: rpc error: code = Unknown desc = Post "https://*****.es.amazonaws.com/_msearch?max_concurrent_shard_requests=5": EOF`

![image](https://github.com/grafana/opensearch-datasource/assets/136579429/de9326df-ca34-4882-bf7b-52b210d54381)


These errors are very rare compared to the `invalid character '}'` ones. So I'm not sure the version could have changed anything. I'll try to look through all the errors later in the day, as I completely missed the EOF ones. Maybe other kinds of errors are hidden :)

**Comment by mhristof** [comment]
> @mhristof Can you also try plugin v2.6.1? There might be a client issue in v2.4.1 that you're running into.

v2.6.1 made my issue go away. I can now see the data in my queries and they are not failing.

**Comment by idastambuk** [comment]
@mhristof thanks for the update! I will close this ticket as it looks like the original issue was fixed with 2.6.1.

@pavriet-boxtal I opened [another issue](https://github.com/grafana/opensearch-datasource/issues/204) to track the problem from your comment, you can follow the progress there and let us know if you find any other errors, as you mentioned.
